### PR TITLE
storage: ban txn cmds when TTL is enabled.

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/27303

Problem Summary:

The TTL feature can break transaction atomicity so we can't run transaction commands with TTL enabled.

### What is changed and how it works?

What's Changed:

Returns an error if runs a transaction command with TTL enabled.

### Check List 

Tests 

- Unit test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -6909,4 +6913,70 @@ mod tests {
...
+    fn test_txn_cmd_with_ttl_enabled() {
+        let storage = TestStorageBuilder::new(DummyLockManager {}, true)
+            .build()
+            .unwrap();
+        let (tx, rx) = channel();
+
+        // TTL commands can pass the check.
+        storage
+            .raw_batch_put(
+                Context::default(),
```
Comment (sticnarf) [inline src/storage/mod.rs:6927]: The `for_cas` field should be enabled if you want it to be a scheduler command.

File: src/storage/mod.rs [hunk 2]
```diff
@@ -6909,4 +6913,70 @@ mod tests {
...
+                "".to_string(),
+                vec![(b"k".to_vec(), b"v".to_vec())],
+                10,
+                expect_ok_callback(tx.clone(), 0),
+            )
+            .unwrap();
+        rx.recv().unwrap();
+        storage
+            .raw_batch_delete(
+                Context::default(),
```
Comment (sticnarf) [inline src/storage/mod.rs:6937]: ditto

File: src/storage/mod.rs (hunk 1 above)
Comment (youjiali1995) [reply src/storage/mod.rs:6927]: I missed  the `_atomic` suffix...

**Review by sticnarf — APPROVED** [review]
LGTM

**Comment by andylokandy** [comment]
I'd rather like to ban txn requests on the service layer because the [API V2](https://github.com/tikv/rfcs/pull/69) is going to do so.

**Comment by youjiali1995** [comment]
> I'd rather like to ban txn requests on the service layer because the [API V2](https://github.com/tikv/rfcs/pull/69) is going to do so.

It's a long term goal. I think we can forbid it in this way for now. @andylokandy

**Comment by andylokandy** [comment]
@youjiali1995 You may get me wrong. I mean we can ban the txn requests in `service.rs` instead of in storage.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Comment by youjiali1995** [comment]
> @youjiali1995 You may get me wrong. I mean we can ban the txn requests in `service.rs` instead of in storage.

@andylokandy Do you mean in `service/kv.rs`, moreover, the `txn_command_future` macro? It's verbose to do it because different response types have different error fields, part of them have `errors` field, other have `error` field.

**Comment by andylokandy** [comment]
@youjiali1995 You can add a light wrapper(a function or a closure) to check the mode before executing the storage command. Mixing the mode check into one specific mode make (txn) will make later life harder.

**Comment by youjiali1995** [comment]
@andylokandy @sticnarf PTAL again.

**Review by sticnarf — COMMENTED** [review]
LGTM

**Review by andylokandy — APPROVED** [review]

**Pull request merged**